### PR TITLE
fixed: null media manager in the unit test environment

### PR DESCRIPTION
**Pull request by malard**
## Description

`CServiceManager::GetMediaManager()` dereferences `m_mediaManager` with no guard, and
`CServiceManager::InitForTesting()` never created it. Any unit test that reached the service
therefore dereferenced a null `unique_ptr` and took the whole test binary down with an access
violation and no stack.

This is reachable from ordinary code rather than being a corner case. Under
`HAS_OPTICAL_DRIVE`, `CDVDFactoryInputStream::CreateInputStream` consults the media manager on
every call:

https://github.com/xbmc/xbmc/blob/master/xbmc/cores/VideoPlayer/DVDInputStreams/DVDFactoryInputStream.cpp#L95

so **no unit test could open a media file at all** — which rules out the demuxer, the codec
factory, `SeekAndDecodeFirstPicture`, `GetFileStreamDetails` and anything built on them.

The failure mode is the awkward part: a bare `0xc0000005` with no message, so the natural
conclusion is "the code I just wrote is broken", not "the test environment is incomplete".

Reduced to a single call to confirm the cause:

```cpp
TEST(Diag, MediaManager)
{
  CServiceBroker::GetMediaManager().TranslateDevicePath("");  // access violation
}
```

Note the addon manager *is* initialised by `InitForTesting`, so the earlier `GetAddonMgr()`
call in the same function is a red herring; the media manager is the first unregistered
service the factory touches.

## Approach

Construct `CMediaManager` in `InitForTesting`, and reset it in `DeinitTesting`.

It is deliberately **not** `Initialize()`d. That call creates a platform storage provider and
enumerates optical drives, which a headless test process has no business doing, and nothing
reachable from the test environment needs it — `TranslateDevicePath`, the call
`CreateInputStream` makes, reads only plain members with safe defaults.

## Effect

The change restores the ability to unit test the whole file-opening path. Verified locally
with a test that previously killed the test binary and now opens and decodes from a real 4K
HEVC file in ~460 ms.

## Tests

`xbmc/test/TestMediaManager.cpp` — asserts the service is available in the test environment,
and that the specific call `CreateInputStream` makes returns rather than crashing.

## Not addressed here

Several other `CServiceManager` accessors have the same shape, returning `*m_member` with no
guard, so any use before the owning init stage is undefined behaviour rather than a
diagnosable error. Making those fail with the name of the missing service would turn a class
of crashes into readable diagnostics, but it is a wider change and is left out to keep this
one to the reported fault. Happy to follow up if that is wanted.

## Checklist

- Builds and passes the test suite locally (MSVC, Windows x64).
- New test file additionally checked with clang 19 on Linux at `-std=c++20 -Wall -Wextra
  -Werror`; no warnings attributable to it.
- No behaviour change outside the unit test environment: `InitForTesting` and `DeinitTesting`
  are not used by the application.

### CODE REVIEW

File: xbmc/test/TestMediaManager.cpp [hunk 1]
```diff
@@ -0,0 +1,40 @@
+/*
+ *  Copyright (C) 2005-2026 Team Kodi
```
Comment (garbear) [inline xbmc/test/TestMediaManager.cpp]: Just 2026 please. All code is written in a context, and 2005 context isn't considered in this file.

**Comment by CrystalP** [comment]
Way too much comment noise / text as is often the case with raw AI-generated code.

There is no problem with the addition of MediaManager to the test environment init/deinit, but as far as I am concerned, should be included as a dedicated commit in the PR that requires it, not as a standalone PR. It would be dead code until then otherwise.

File: xbmc/ServiceManager.cpp [hunk 2]
```diff
@@ -87,6 +87,16 @@ bool CServiceManager::InitForTesting()
...
 
+  // Needed because CServiceManager::GetMediaManager() dereferences this unconditionally,
+  // so anything reaching it without one - CDVDFactoryInputStream::CreateInputStream, for
+  // instance - takes the whole test binary down with an access violation rather than
+  // failing in a way that can be diagnosed.
+  //
+  // Deliberately constructed but not Initialize()d: that creates a platform storage
+  // provider and enumerates optical drives, which a headless test process has no business
+  // doing. Everything reachable from the test environment reads only plain members.
+  m_mediaManager = std::make_unique<CMediaManager>();
```
Comment (garbear) [inline xbmc/ServiceManager.cpp]: IMO drop the entire comment, dereferencing members unconditionally is the pattern of the file, and it's obvious that objects constructed here are for headless testing. If members beyond plain ones were used, tests would fail.

**Comment by garbear** [comment]
Fix is correct, but @CrystalP is right, maybe this should be introduced in the PR where it's first needed.

**Comment by malard** [comment]
the PR I need it in, will fall in the next day or two, I am building that PR ontop of this PR

**Comment by malard** [comment]
agreed on the junk comments, sorry thats sloppy on my part

File: xbmc/test/TestMediaManager.cpp (hunk 1 above)
Comment (malard) [reply xbmc/test/TestMediaManager.cpp]: Fixed in 761666bd — header is now 2026 only.

File: xbmc/ServiceManager.cpp (hunk 2 above)
Comment (malard) [reply xbmc/ServiceManager.cpp]: Agreed, dropped in 761666bd. The construction now stands alone with no comment.

I applied the same reasoning to `TestMediaManager.cpp` and removed the block comment there too — it narrated the bug rather than justifying the code, and that history is already in the commit message and the PR description. Also replaced a tautological `EXPECT_TRUE(s.empty() || !s.empty())` with `EXPECT_NO_FATAL_FAILURE`, which says what the test actually checks.

**Comment by garbear** [comment]
What about

```c++
CServiceManager::CServiceManager()
  : m_fileExtensionProvider(std::make_unique<CFileExtensionProvider>()),
    m_mediaManager(std::make_unique<CMediaManager>())
{
}
```

But the commit here is great. I'd include it as the first commit in the PR that requires it, so that code spelunking in the future can be more direct.

**Comment by malard** [comment]
Thanks — adopted your suggestion, and it turned out better than the original: constructing it in the initializer list means the `InitForTesting()` construction and the matching `DeinitTesting()` reset both disappear, so the commit is now a net two lines in `ServiceManager.cpp`. `Initialize()` stays in stage two for the reasons in the commit message, and the application path is untouched.

Closing this per your other point. The commit will reappear as the first commit of the content-geometry PR, which is the work that needs it — that PR is not open yet, there are several slices still to land, so this is going quiet rather than disappearing.

**Comment by CrystalP** [comment]
> What about
> 
> ```c++
> CServiceManager::CServiceManager()
>   : m_fileExtensionProvider(std::make_unique<CFileExtensionProvider>()),
>     m_mediaManager(std::make_unique<CMediaManager>())
> {
> }
> ```

Not sure about that. That creates an exception to the pattern - save for the extension provider, don't know what the thinking was there.
Isn't the point of the service manager is to bring services up and down in a controlled order - can you guarantee that creating an instance does not / will not depend on something else?

How does it save you from the reset in Deinit? The established pattern is very clear as well and ensures a controlled destruction order.

There could be a PR to improve the way services are brought up and down but that's probably not the purpose of your planned follow up PR.

**Comment by garbear** [comment]
> Isn't the point of the service manager is to bring services up and down in a controlled order

Constructors shouldn't have side effects, so service bringup should be in an init() function. That function would go where the make_unique() calls currently are.

> How does it save you from the reset in Deinit? 

Deinit() must not call reset(). Nothing should. Service bringdown, if explicitly required, should happen in the appropriate place with a deinit function in the service class.

> can you guarantee that creating an instance does not / will not depend on something else?

If instance creation depends on another service, this approach would need care. But the dependent service should probably be passed in the constructor so that it's always available to media manager. In that case, we can see no dependencies passed to the constructor here.

IMO the constructor approach is preferred when possible, because allocation should happen early on (as the deference accessors are always available), and initialization should happen separately where appropriate.
